Update topic and stream headers on update message events mid-compose.

prah23 opened the pull request:
**PR structure**:
- **tests: Generalize the update message index using factory.**
This commit adds a factory to generate the messages index relevant to
update message event tests, in an effort to reduce code duplications
for future tests. The factory requires three parameters - msg_stream_id,
(stream_id of the messages in the index), topics (dict containing streams
as keys, and a list of topics as the values, consistent with the structure
in the index) and message_ids (which is a list of the message ids to be
generated in the index). The factory uses these parameters to customize
the index.
- **boxes: Create helper function to update the topic compose header.**
This commit adds a helper function facilitating abstracted updates
to the topic in the topic compose header (title_write_box), and also
sets the edit_pos appropriately.
Test added.
- **api_types: Extend UpdateMessageEvent to accomodate topic updates.**
This commit includes more fields under the UpdateMessageEvent class
to ensure type-consistent handling for said fields in the update message
events conveying topic updates.
- **model: Handle topic updates in the compose header mid-compose.**
This commit enables dynamic behaviour in the compose box when a topic
is updated when the user is currently composing. The topic that the user
is currently writing a message to is checked and the relevant update is
done if the topic in the event matches the corresponding header.
Tests added.
- **boxes: Create helper function to update the stream compose header.**
This commit adds a helper function facilitating abstracted updates
to the stream name in the stream compose header (title_write_box),
the corresponding edit_pos modification, as well as the stream_id
attribute of the WriteBox object in use.
Test added.
- **api_types: Extend UpdateMessageEvent to handle cross-stream updates.**
This commit extends the UpdateMessageEvent class by including the
new_stream_id field to ensure type-consistent handling of cross-stream
updates (moving topics between streams) in the update message events.
- **model: Handle stream updates in the compose header mid-compose.**
This commit updates the compose header when a stream is updated while
the user is currently composing. The stream that the user is currently
writing a message to is checked and the relevant update is done if the
stream_id in the event matches the corresponding header.
Tests added.

**Testing and linting:**
I've ran tests locally on each commit. I've also ran `black` checks and then the more extensive `./tools/lint-all`.

Fixes #1040.

prah23 committed:
tests: Generalize the update message index using factory.

This commit adds a factory to generate the messages index relevant to
update message event tests, in an effort to reduce code duplications
for future tests. The factory requires three parameters - msg_stream_id,
(stream_id of the messages in the index), topics (dict containing streams
as keys, and a list of topics as the values, consistent with the structure
in the index) and message_ids (which is a list of the message ids to be
generated in the index). The factory uses these parameters to customize
the index.

neiljp reviewed (commented):
@prah23 Generally this looks good, though I've not read the first commit, and I think it might be clearer to focus on updating different parts of the compose area in different commits, as you use certain api fields and helper functions for the different features?

prah23 commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -314,6 +314,17 @@ def _set_stream_write_box_style(self, widget: ReadlineEdit, new_text: str) -> No
             (color, stream_marker)
         )
 
+    def _update_title_write_box(self, new_topic: str) -> None:
+        self.title_write_box.set_edit_text(new_topic)
+        self.title_write_box.set_edit_pos(len(new_topic))
+
+    def _update_stream_write_box(
```
Right, I was a little conflicted about that myself. Understood, I'll rename them.

prah23 replied:
Right, I was a little conflicted about that myself. Understood, I'll rename them.

prah23 commented on tests/model/test_model.py:
```diff
@@ -1342,6 +1342,7 @@ def _factory():
                     "subject": "new subject",
                     "stream_id": 1,
                     "message_ids": [1],
+                    "propagate_mode": "change_one",
```
No, it doesn't but I check for the `propagate_mode` key's value in the model, and hence the tests fail without it. It's certain that the `propagate_mode` key exists if the `subject` key included in the event.

neiljp replied:
Right, so this can go into the commit that adds this feature - topic updates?

prah23 replied:
Yes, this is currently a part of the final commit (topic and stream update features combined). I'll break that commit down put these changes up in the topic updates commit.

prah23 committed:
boxes: Create helper function to update the topic compose header.

This commit adds a helper function facilitating abstracted updates
to the topic in the topic compose header (title_write_box), sets
the edit_pos appropriately and also notifies the user of this topic
change using `report_warning()`.

Test added.

prah23 committed:
boxes: Create helper function to update the stream compose header.

This commit adds a helper function facilitating abstracted updates
to the stream name in the stream compose header (title_write_box),
the corresponding edit_pos modification, as well as the stream_id
attribute of the WriteBox object in use. The helper also notifies
the user of the topic being moved to another stream using
`report_warning()`.

Test added.

prah23 committed:
model: Handle topic updates in the compose header mid-compose.

This commit enables dynamic behaviour in the compose box when a topic
is updated when the user is currently composing. The topic that the user
is currently writing a message to is checked and the relevant update is
done if the topic in the event matches the corresponding header.

Tests added.

prah23 committed:
model: Handle stream updates in the compose header mid-compose.

This commit updates the compose header when a stream is updated while
the user is currently composing. The stream that the user is currently
writing a message to is checked and the relevant update is done if the
stream_id in the event matches the corresponding header.

Tests added.

Fixes #1040.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -314,6 +314,10 @@ def _set_stream_write_box_style(self, widget: ReadlineEdit, new_text: str) -> No
 
+    def update_topic_compose_header(self, new_topic: str) -> None:
+        self.title_write_box.set_edit_text(new_topic)
+        self.title_write_box.set_edit_pos(len(new_topic))
```
Did you want to assert on positions too, generally?

prah23 replied:
Sounds good, I'll update accordingly.

neiljp reviewed (commented):
@prah23 Good work adjusting the commits - these look simpler and more independent of each other :+1:
That makes them easier to read, review, and potentially reorder and squash/adjust (slightly!) further...

For commit structure we could now move the helper function commits just before each related new feature (event) commit. Generally if I see a commit that doesn't have any obvious connections (doesn't connect to other code), I expect its use will become obvious next/soon in the commit flow - so one benefit is we shouldn't generally need to note that it's a prep-commit in the commit text (refer to earlier/later commits generally) unless we have a large set of commits that can't be untangled/simplified. It also has the benefit that we don't need to adjust focus to thinking about "some potentially much later time" when reading commits based on the commit text, but just the code :)
Now that those feature commits are separate, we can also split that small api_types commit into two and add to the feature commits where they're first used. Improving the typing is good, but keeping commits "atomic" where possible is good too.
(we could include the helpers too, but then we end up with extra functions + tests per commit again!)

If you adjust like this, then compared to your original commits you should be able to see the two independent features can be read (even perhaps merged) individually, and are easier to review to enable that too :)

prah23 commented on tests/model/test_model.py:
```diff
@@ -1617,6 +1624,57 @@ def _set_topics_to_old_and_new(event):
+        model._handle_update_message_event(event)
+
+        if (
+            compose_box_open
+            and propagate_mode in ["change_later", "change_all"]
+            and original_topic == event["orig_subject"]
+            and original_stream_id == event["stream_id"]
```
I'm not sure what alternative strategy here would be effective, do you have any thoughts on this?

neiljp replied:
See the stream for discussion.

prah23 commented on zulipterminal/model.py:
```diff
@@ -1206,6 +1206,20 @@ def _handle_update_message_event(self, event: Event) -> None:
+            # i.e., a 'change_later' or 'change_all' change in the topic.
+            # The logic here has been inspired by the web app's
+            # `static/js/message_events.js`.
+            if event["propagate_mode"] in ["change_later", "change_all"]:
+                if hasattr(self.controller, "view"):
+                    view = self.controller.view
+                    old_subject = event["orig_subject"]
+                    if view.write_box.stream_id == stream_id:
+                        if view.write_box.title_write_box.edit_text == old_subject:
+                            view.write_box.update_topic_compose_header(new_subject)
```
Do you mean extending the "alias" further to `self.controller.view.write_box` and giving that a name instead? So that it makes the code below more compact.

neiljp replied:
That's what I meant by the last sentence, yes.

prah23 commented on tests/model/test_model.py:
```diff
@@ -1310,34 +1310,57 @@ def test_notify_users_enabled(
+                    for message_id in [1, 2]
+                },
+                "edited_messages": set(),
+                "topics": {1: ["old subject", "other subject"], 2: []},
```
Inconsistent as in? I'm not very confident about the logic here, but I've added `other_subject` only where `old subject` is already present in the `dict` in the cases from the test below.

neiljp replied:
Two topics in stream id `1`?

prah23 commented on tests/model/test_model.py:
```diff
@@ -1617,6 +1624,57 @@ def _set_topics_to_old_and_new(event):
+        view = model.controller.view
+        view.write_box.update_topic_compose_header = mocker.Mock()
+
+        if compose_box_open:
+            view.write_box.stream_id = original_stream_id
+            view.write_box.title_write_box.edit_text = original_topic
```
When it's False, the stream_id is `None` and the `edit_text` is an empty string. You're suggesting that I mention that to portray a closed compose box here, yes?

neiljp replied:
In the test this is a mock, so do those values matter? If so, you need to set them.

Regarding `compose_box_open` in general, this seems to work but it's a little confusing as a shortcut for... well is it just testing those values? Or is it intended as setup, which then later appears in the assert selection? It just feels suspicious for some reason

prah23 committed:
api_types: Extend UpdateMessageEvent to accomodate topic updates.

This commit includes more fields under the UpdateMessageEvent class
to ensure type-consistent handling for said fields in the update message
events conveying topic updates.

prah23 committed:
api_types: Extend UpdateMessageEvent to handle cross-stream updates.

This commit extends the UpdateMessageEvent class by including the
new_stream_id field to ensure type-consistent handling of cross-stream
updates (moving topics between streams) in the update message events.

prah23 commented on zulipterminal/model.py:
```diff
@@ -1205,6 +1205,36 @@ def _handle_update_message_event(self, event: Event) -> None:
                     else:
                         self.index["topics"][stream_id] = []
 
+            if event["propagate_mode"] in ["change_later", "change_all"]:
+                old_subject = event["orig_subject"]
+
+                # If the user is composing a message to this topic, update the topic in
+                # the title_write_box header on a 'change_later' or 'change_all' update
+                # in the topic.
+                if hasattr(self.controller, "view"):
+                    write_box = self.controller.view.write_box
+                    if (
+                        write_box.stream_id == stream_id
+                        and write_box.title_write_box.edit_text == old_subject
+                    ):
+                        write_box.update_topic_compose_header(new_subject)
+
```
I've rewritten this code in a way that facilitates a smooth extension when adding support for moving topics between streams, which is why it is a little different from the previous iteration.

prah23 commented:
Please hold on the stream headers update aspect of this PR for now, the topic header update is ready.

neiljp commented on tests/model/test_model.py:
```diff
@@ -1757,6 +1764,66 @@ def _set_topics_to_old_and_new(event):
+        }
+
+        model.index = update_message_event_index_factory(
+            msg_stream_id=1,
+            topics={1: ["old subject", "other subject"], 2: []},
+            message_ids=[1, 2],
```
Just glancing at this alone, it's not as obvious as it could be what this translates to:
* topics -> topics_per_stream_id?
* the message_ids and msg_stream_id are coupled? If so, I'd prefer to see the data coupled and not repeated
* what are the message_ids in? If it's a default name, then we don't need to specify it, perhaps...

...or ultimately, maybe just one nested structure could work?
`{stream_id: {"old_subject": [1, 2], "other_subject": []}`

Not all of that may be necessary to get this PR bootstrapped, but it's definitely the case that we should avoid reusing numbers for streams and message ids to avoid confusion :)

neiljp commented:
@prah23 Heya - remember this PR? It's pretty useful and it'd be great to get at least partially merged :)

I remembered this PR was still open and have just rebased it such that in principle it passes tests & linting - I've not pushed it here, as I wanted to check with you first to make sure you have a version available for comparison if you want?

The changes required were:
- adjusting the test factory function; I think it may have been a PR/commit of yours that added an additional model parameter. Provisionally I added this to the parameters, but as per my previous comment on the PR this may be able to be simplified.
- added test function types; again, that was your doing in another PR, but its easy enough to change :)
- one event type change is dropped, and the second for stream changes is still there but may be a small change incorporated into another commit if #1401 merges first

We have `tools/check-branch` now, which is pretty useful for fixing these kinds of changes :)
(as well as checking - ensuring commits pass individually in CI)

I've only tested this for topic moves (I'm not sure if the stream part was complete) and this works as before from what I remember :+1:

Outstanding points are:
- this is not necessarily related to this issue, but we have a bug which relates to if the topic list is updated then we get an error relating to the search_lock overwriting the screen; that doesn't block the topic update, but it messes up the UI and it's difficult to tell what's happening :/
- if there is only one message in the topic and only it changes name, should this change a response, do you think?
- we now have a single draft saved; I've not checked with the web app for comparison, but I wonder if we should update any draft topics based on events too?
- do you remember the status of the stream updating part?

I'll do a deeper review before moving on or merging (partial or otherwise) in any case.
